File: Day<30>/Solution.py
class Solution:
    def build_bridges_util(self,str1, str2, i, j, dp):
        if i >= len(str1) or j >= len(str2):
            return 0
    
        if dp[i][j] != -1:
            return dp[i][j]
    
    
        if str1[i] == str2[j]:
            dp[i][j] = 1 + self.build_bridges_util(str1, str2, i + 1, j + 1, dp)
            return dp[i][j]
    
        dp[i][j] = max(self.build_bridges_util(str1, str2, i + 1, j, dp), self.build_bridges_util(str1, str2, i, j + 1, dp))
        
    
        return dp[i][j]
    
    
    def build_bridges(self, str1, str2):
        n1, n2 = len(str1), len(str2)
        dp = [[-1 for i in range(n2 + 1)] for j in range(n1 + 1)]
    
        return self.build_bridges_util(str1, str2, 0, 0,dp)


#{ 
#  Driver Code Starts
if __name__ == '__main__':
    T = int(input())
    for _ in range(T):
        obj = Solution()
        str1, str2 = input().split()
        print(obj.build_bridges(str1, str2))

File: Day<30>/Solution.cpp
#include<bits/stdc++.h>
using namespace std; 


 // } Driver Code Ends
class Solution
{
    public:
    int build_bridges_util(int i , int j , string str1 , string str2 , int n , int m , int dp[105][105])
    {
        if(i>=n or j>=m)
            return 0;
        
        int ans = dp[i][j];
        if(ans!=-1)
            return ans;
        
        ans = 0;
        if(str1[i]==str2[j])
            ans = 1 + build_bridges_util(i+1, j+1, str1, str2,n, m, dp);
        
        ans = max(ans, build_bridges_util(i+1, j, str1, str2, n, m, dp));
        ans = max(ans, build_bridges_util(i, j+1, str1, str2, n, m, dp));
        
        return dp[i][j] = ans;
    }
    
    int build_bridges(string str1, string str2)
    {
        int dp[105][105];
        memset(dp, -1, sizeof dp);
        int n = str1.size();
        int m = str2.size();
        return build_bridges_util(0,0,str1,str2,n,m,dp);
    }
};

// { Driver Code Starts.
int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(NULL);
    cout.tie(NULL);

    int t;
    cin >> t;

    while(t--)
    {
        string str1,str2;
        cin >> str1 >> str2;
        Solution obj;
        cout<<obj.build_bridges(str1,str2)<<endl;
    } 
    return 0;
}
